File: Resources/goagent-server/python/fetch.py
#!/usr/bin/env python
# coding=utf-8
# Based on GAppProxy by Du XiaoGang <dugang@188.com>
# Based on WallProxy 0.4.0 by hexieshe <www.ehust@gmail.com>

__version__ = '1.8.2'
__author__ =  'phus.lu@gmail.com'
__password__ = ''

import sys, os, re, time, struct, zlib, binascii, logging
import webapp2
from google.appengine.api import urlfetch
from google.appengine.runtime import apiproxy_errors, DeadlineExceededError

FetchMax = 3
FetchMaxSize = 1024*1024*4
Deadline = 30

def encode_data(dic):
    return '&'.join('%s=%s' % (k, binascii.b2a_hex(v)) for k, v in dic.iteritems() if v)

def decode_data(qs):
    return dict((k, binascii.a2b_hex(v)) for k, _, v in (x.partition('=') for x in qs.split('&')))

class MainPage(webapp2.RequestHandler):

    def send_response(self, status, headers, content):
        strheaders = encode_data(headers)
        #logging.debug('response status=%s, headers=%s, content length=%d', status, headers, len(content))
        if headers.get('content-type', '').startswith(('text/', 'application/json', 'application/javascript')):
            data = '1' + zlib.compress('%s%s%s' % (struct.pack('>3I', status, len(strheaders), len(content)), strheaders, content))
        else:
            data = '0%s%s%s' % (struct.pack('>3I', status, len(strheaders), len(content)), strheaders, content)
        self.response.headers['Content-Type'] = 'image/gif'
        self.response.out.write(data)

    def send_notify(self, method, url, status, content):
        logging.warning('%r Failed: url=%r, status=%r', method, url, status)
        content = '<h2>Python Server Fetch Info</h2><hr noshade="noshade"><p>%s %r</p><p>Return Code: %d</p><p>Message: %s</p>' % (method, url, status, content)
        self.send_response(status, {'content-type':'text/html'}, content)

    def post(self):
        request = decode_data(zlib.decompress(self.request.body))
        #logging.debug('post() get fetch request %s', request)

        method = request['method']
        url = request['url']
        payload = request['payload']

        if __password__ and __password__ != request.get('password', ''):
            return self.send_notify(method, url, 403, 'Wrong password.')

        fetchmethod = getattr(urlfetch, method, '')
        if not fetchmethod:
            return self.send_notify(method, url, 501, 'Invalid Method')

        if 'http' != url[:4]:
            return self.send_notify(method, url, 501, 'Unsupported Scheme')

        deadline = Deadline

        headers = dict((k.title(), v.lstrip()) for k, _, v in (line.partition(':') for line in request['headers'].splitlines()))
        headers['Connection'] = 'close'

        errors = []
        for i in xrange(int(request.get('fetchmax', FetchMax))):
            try:
                response = urlfetch.fetch(url, payload, fetchmethod, headers, False, False, deadline, False)
                break
            except apiproxy_errors.OverQuotaError, e:
                time.sleep(4)
            except DeadlineExceededError, e:
                errors.append(str(e))
                logging.error('DeadlineExceededError(deadline=%s, url=%r)', deadline, url)
                time.sleep(1)
                deadline = Deadline * 2
            except urlfetch.DownloadError, e:
                errors.append(str(e))
                logging.error('DownloadError(deadline=%s, url=%r)', deadline, url)
                time.sleep(1)
                deadline = Deadline * 2
            except urlfetch.InvalidURLError, e:
                return self.send_notify(method, url, 501, 'Invalid URL: %s' % e)
            except urlfetch.ResponseTooLargeError, e:
                response = e.response
                logging.error('DownloadError(deadline=%s, url=%r) response(%s)', deadline, url, response and response.headers)
                if response and response.headers.get('content-length'):
                    response.status_code = 206
                    response.headers['accept-ranges']  = 'bytes'
                    response.headers['content-range']  = 'bytes 0-%d/%s' % (len(response.content)-1, response.headers['content-length'])
                    response.headers['content-length'] = len(response.content)
                    break
                else:
                    headers['Range'] = 'bytes=0-%d' % FetchMaxSize
                deadline = Deadline * 2
            except Exception, e:
                errors.append(str(e))
                if i==0 and method=='GET':
                    deadline = Deadline * 2
        else:
            return self.send_notify(method, url, 500, 'Python Server: Urlfetch error: %s' % errors)

        headers = response.headers
        if 'set-cookie' in headers:
            scs = headers['set-cookie'].split(', ')
            cookies = []
            i = -1
            for sc in scs:
                if re.match(r'[^ =]+ ', sc):
                    try:
                        cookies[i] = '%s, %s' % (cookies[i], sc)
                    except IndexError:
                        pass
                else:
                    cookies.append(sc)
                    i += 1
            headers['set-cookie'] = '\r\nSet-Cookie: '.join(cookies)
        headers['connection'] = 'close'
        return self.send_response(response.status_code, headers, response.content)

    def get(self):
        html = u'''\
<html>
<head>
    <link rel="icon" type="image/vnd.microsoft.icon" href="http://www.google.cn/favicon.ico">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>GoAgent Python Server %(version)s 已经在工作了</title>
</head>
<body>
    <table width="800" border="0" align="center">
        <tr><td align="center"><hr></td></tr>
        <tr><td align="center">
            <b><h1>GoAgent Python Server %(version)s 已经在工作了</h1></b>
        </td></tr>
        <tr><td align="center"><hr></td></tr>

        <tr><td align="center">
            GoAgent是一个开源的HTTP Proxy软件,使用Python编写,运行于Google App Engine平台上.
        </td></tr>
        <tr><td align="center"><hr></td></tr>

        <tr><td align="center">
            更多相关介绍,请参考<a href="http://code.google.com/p/goagent/">GoAgent项目主页</a>.
        </td></tr>
        <tr><td align="center"><hr></td></tr>

    </table>
</body>
</html>
''' % dict(version=__version__)
        self.response.headers['Content-Type'] = 'text/html; charset=utf-8'
        self.response.out.write(html.encode('utf8'))

app = webapp2.WSGIApplication([('/fetch.py', MainPage)])

File: Resources/goagent-server/golang/fetch/fetch.go
// Copyright 2012 Phus Lu. All rights reserved.

package fetch

import (
    "fmt"
	"bytes"
	"strings"
	"strconv"
	"regexp"
	"time"
    "io/ioutil"
	"compress/zlib"
	"encoding/binary"
	"encoding/hex"
	"net/http"

	"appengine"
	"appengine/urlfetch"
)

const (
	Version  = "1.8.2"
	Author   = "phus.lu@gmail.com"
	Password = ""

	FetchMax     = 3
	FetchMaxSize = 1024 * 1024 * 4
	Deadline int64 = 30 *1e9
)

func encodeData(dic map[string]string) []byte {
	w := bytes.NewBufferString("")
	for k, v := range dic {
	    if len(v) != 0 {
		    fmt.Fprintf(w, "&%s=%s", k, hex.EncodeToString([]byte(v)))
		}
	}
	return w.Bytes()[1:]
}

func decodeData(qs []byte) map[string]string {
	m := make(map[string]string)
	for _, kv := range strings.Split(string(qs), "&") {
		if kv != "" {
			pair := strings.Split(kv, "=")
			value, _ := hex.DecodeString(pair[1])
			m[pair[0]] = string(value)
		}
	}
	return m
}

type Handler struct {
	response http.ResponseWriter
	request  *http.Request
	context  appengine.Context
	http.Handler
}

func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	h.response = w
	h.request = r
	h.context = appengine.NewContext(r)
	if r.Method == "POST" {
		h.post()
	} else {
		h.get()
	}
}

func (h Handler) printResponse(status int, header map[string]string, content []byte) {
	headerEncoded := encodeData(header)

	h.response.WriteHeader(200)
	h.response.Header().Set("Content-Type", "image/gif")

    compressed := false
	if contentType, ok := header["content-type"]; ok {
	    if strings.HasPrefix(contentType, "text/") || strings.HasPrefix(contentType, "application/json") || strings.HasPrefix(contentType, "application/javascript") {
	        compressed = true
	    }
	}

	if compressed {
		h.response.Write([]byte("1"))
		w, err := zlib.NewWriterLevel(h.response, zlib.BestCompression)
		if err != nil {
			h.context.Criticalf("zlib.NewWriterLevel(h.response, zlib.BestCompression) Error: %v", err)
			return
		}
		defer w.Close()
		binary.Write(w, binary.BigEndian, uint32(status))
		binary.Write(w, binary.BigEndian, uint32(len(headerEncoded)))
		binary.Write(w, binary.BigEndian, uint32(len(content)))
		w.Write(headerEncoded)
		w.Write(content)
	} else {
		h.response.Write([]byte("0"))
		binary.Write(h.response, binary.BigEndian, uint32(status))
		binary.Write(h.response, binary.BigEndian, uint32(len(headerEncoded)))
		binary.Write(h.response, binary.BigEndian, uint32(len(content)))
		h.response.Write(headerEncoded)
		h.response.Write(content)
	}
}

func (h Handler) printNotify(method string, url string, status int, text string) {
	content := []byte(fmt.Sprintf("<h2>Go Server Fetch Info</h2><hr noshade='noshade'><p>%s '%s'</p><p>Return Code: %d</p><p>Message: %s</p>", method, url, status, text))
	headers := map[string]string{"content-type": "text/html"}
	h.printResponse(status, headers, content)
}

func (h Handler) post() {
	r, err := zlib.NewReader(h.request.Body)
	if err != nil {
		h.context.Criticalf("zlib.NewReader(h.request.Body) Error: %v", err)
		return
	}
	defer r.Close()
	data, err := ioutil.ReadAll(r)
	if err != nil {
		h.context.Criticalf("ioutil.ReadAll(r) Error: %v", err)
		return
	}
	request := decodeData(data)

	method := request["method"]
	url := request["url"]
	headers := request["headers"]

	if Password != "" {
		password, ok := request["password"]
		if !ok || password != Password {
			h.printNotify(method, url, 403, " Wrong Password.")
		}
	}

	if !strings.HasPrefix(url, "http") {
		h.printNotify(method, url, 501, "Unsupported Scheme")
	}

	payload := request["payload"]
	req, err := http.NewRequest(method, url, bytes.NewBufferString(payload))
	if err != nil {
		h.printNotify(method, url, 500, "http.NewRequest(method, url, payload) failed")
	}

	for _, line := range strings.Split(headers, "\r\n") {
		kv := strings.SplitN(line, ":", 2)
		if len(kv) == 2 {
		    req.Header.Set(strings.Title(kv[0]), strings.TrimSpace(kv[1]))
		}
	}

	fetchmax := FetchMax
	if fetchmaxString, ok := request["fetchmax"] ; ok {
	    fetchmax, err = strconv.Atoi(fetchmaxString)
	    if err != nil {
	        h.context.Errorf("strconv.Atoi(fetchmaxString=%v) error=%v", fetchmaxString, err)
	        fetchmax = FetchMax
	    }
	}
	
	deadline := time.Duration(Deadline)
	
	var errors []string
	for i := 0; i < fetchmax; i++ {
		t := &urlfetch.Transport{Context:h.context, Deadline:deadline, AllowInvalidServerCertificate:true}
		resp, err := t.RoundTrip(req)
		if err != nil {
			message := err.Error()
			errors = append(errors, message)
			if strings.Contains(message, "FETCH_ERROR") {
				h.context.Errorf("URLFetchServiceError_FETCH_ERROR(type=%T, deadline=%v, url=%v)", err, deadline, url)
				time.Sleep(1*1e9)
				deadline = time.Duration(Deadline*2)
			} else if strings.Contains(message, "DEADLINE_EXCEEDED") {
				h.context.Errorf("URLFetchServiceError_DEADLINE_EXCEEDED(type=%T, deadline=%v, url=%v)", err, deadline, url)
				time.Sleep(1*1e9)
				deadline = time.Duration(Deadline*2)
			} else if strings.Contains(message, "INVALID_URL") {
				h.printNotify(method, url, 501, fmt.Sprintf("Invalid URL: %s", err.Error()))
				return
			} else if strings.Contains(message, "RESPONSE_TOO_LARGE") {
				h.context.Errorf("URLFetchServiceError_RESPONSE_TOO_LARGE(type=%T, deadline=%v, url=%v)", err, deadline, url)
				req.Header.Set("Range", fmt.Sprintf("bytes=0-%d", FetchMaxSize))
				//h.context.Infof("req.Header=%v", req.Header)
				deadline = time.Duration(Deadline*2)
			} else {
				h.context.Errorf("URLFetchServiceError UNKOWN(type=%T, deadline=%v, url=%v, error=%v)", err, deadline, url, err)
				time.Sleep(4*1e9)
			}
			continue
		}

		status := resp.StatusCode
		header := make(map[string]string)
		for k, vv := range resp.Header {
		    key := strings.ToLower(k)
			if key != "set-cookie" {
				header[key] = vv[0]
			} else {
				var cookies []string
				i := -1
				regex, _ := regexp.Compile("^[^ =]+ ")
				for _, sc := range strings.Split(vv[0], ", ") {
					if 0 <= i && regex.MatchString(sc) {
						cookies[i] = fmt.Sprintf("%s, %s", cookies[i], sc)
					} else {
						cookies = append(cookies, sc)
						i++
					}
				}
				header["set-cookie"] = strings.Join(cookies, "\r\nSet-Cookie: ")
			}
		}

		content, err := ioutil.ReadAll(resp.Body)
		if err == urlfetch.ErrTruncatedBody {
			h.context.Criticalf("ioutil.ReadAll(resp.Body) return urlfetch.ErrTruncatedBody")
		}
		if status == 206 {
			header["accept-ranges"] = "bytes"
			header["content-length"] = strconv.Itoa(len(content))
		}
		header["connection"] = "close"

		//h.printNotify(method, url, 502, fmt.Sprintf("status=%d, header=%v, len(content)=%d", status, resp.Header, len(content)))
		h.printResponse(status, header, content)
		return
	}
	h.printNotify(method, url, 502, fmt.Sprintf("Go Server Fetch Failed: %v", errors))
}

func (h Handler) get() {
	h.response.WriteHeader(http.StatusOK)
	h.response.Header().Set("Content-Type", "text/html; charset=utf-8")
	fmt.Fprintf(h.response, `
<html>
<head>
    <link rel="icon" type="image/vnd.microsoft.icon" href="http://www.google.cn/favicon.ico">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>GoAgent GAE/Go %s &#x5DF2;&#x7ECF;&#x5728;&#x5DE5;&#x4F5C;&#x4E86;</title>
</head>
<body>
    <table width="800" border="0" align="center">
        <tr><td align="center"><hr></td></tr>
        <tr><td align="center">
            <b><h1>GoAgent Go Server %s &#x5DF2;&#x7ECF;&#x5728;&#x5DE5;&#x4F5C;&#x4E86;</h1></b>
        </td></tr>
        <tr><td align="center"><hr></td></tr>

        <tr><td align="center">
            GoAgent&#x662F;&#x4E00;&#x4E2A;&#x5F00;&#x6E90;&#x7684;HTTP Proxy&#x8F6F;&#x4EF6;, &#x4F7F;&#x7528;Go/Python&#x7F16;&#x5199;, &#x8FD0;&#x884C;&#x4E8E;Google App Engine&#x5E73;&#x53F0;&#x4E0A;
        </td></tr>
        <tr><td align="center"><hr></td></tr>

        <tr><td align="center">
            &#x66F4;&#x591A;&#x76F8;&#x5173;&#x4ECB;&#x7ECD;, &#x8BF7;&#x53C2;&#x8003;<a href="http://code.google.com/p/goagent/">GoAgent&#x9879;&#x76EE;&#x4E3B;&#x9875;</a>.
        </td></tr>
        <tr><td align="center"><hr></td></tr>

    </table>
</body>
</html>`, Version, Version)
}

func init() {
	http.Handle("/fetch.py", Handler{})
}
